tests/postgres_tests/test_array.py

```diff
@@ -1339,6 +1339,22 @@ def test_invalid_char_length(self):
             ],
         )
 
+    def test_invalid_char_length_with_remove_trailing_nulls(self):
+        field = SplitArrayField(
+            forms.CharField(max_length=2, required=False),
+            size=3,
+            remove_trailing_nulls=True,
+        )
+        with self.assertRaises(exceptions.ValidationError) as cm:
+            field.clean(["abc", "", ""])
+        self.assertEqual(
+            cm.exception.messages,
+            [
+                "Item 1 in the array did not validate: Ensure this value has at most 2 "
+                "characters (it has 3).",
+            ],
+        )
+
```

Comment:
```suggestion

    def test_invalid_char_length_with_remove_trailing_nulls_with_form(self):
        class SplitArrayForm(forms.Form):
            field = SplitArrayField(
                forms.CharField(max_length=3, required=True),
                size=3,
                remove_trailing_nulls=True,
                required=False,
            )

        form = SplitArrayForm(data={'field': ["abc", "", ""]})
        self.assertEqual(form.is_valid(), False)
```
This test (which is more similar to what the ticket author shared) still fails. I don't know why - can you have a look?

Reply:
For the test example that Sarah provided above, it seems like the value being passed into the clean function of [SplitArrayForm](https://github.com/django/django/blob/5ed72087c450af1a5da138bdfa674a069cf3f09c/django/contrib/postgres/forms/array.py#L212) is `[None, None, None]` when it is expected to be `["abc", "", ""]`. I haven't looked further to see why this is the case, but wanted to leave this observation here.

Reply:
Also @sarahboyce if we're testing invalid char length, shouldn't the max_length=2 in this case?

Reply:
Hi there @ontowhee thanks for the observation, could you teach me how you could trace variables through such a complex project?

Reply:
> Also @sarahboyce if we're testing invalid char length, shouldn't the max_length=2 in this case?

Yes you're right ✅️

Reply:
I think the reason why the value being passed into the clean function is `[None, None, None]` is because SplitArrayField expects data where each element is a separate form field, when data is provided as `{"field": ["abc", "", ""]}`, Django's form processing doesn't recognise this as data for a SplitArrayField. Instead, it treats it as a single field named "field" with a list value. The SplitArrayField then doesn't know how to properly bind this data to its subfields. As a result, `form["field"].data` will be `[None, None, None]`. However, if data is provided as `{"field_0": "abc", "field_1": "", "field_2": ""}` then the form correctly recognises this and `form["field"].data` will be `['abc', '', '']` and `form.is_valid()` will return `False`. I could be wrong though

Reply:
@GappleBee I think you're right about the data being passed into the form. Looking at the other tests, like [test_remove_trailing_nulls](https://github.com/django/django/blob/5ed72087c450af1a5da138bdfa674a069cf3f09c/tests/postgres_tests/test_array.py#L1257), the data provided is in the format you described.

I ran the test with `data = {"field_0": "abc", "field_1": "", "field_2": ""}`. However, `form.cleaned_data` is empty when it is expected to be `{"field": ["abc"]}` (this is after it removes trailing nulls). Can you take a look here?

Reply:
I would say it's because the `field_0` in your case stores a string that's too long (more than two characters long) and so it raises a `ValidationError`. Therefore, instead of assigning a value to form.cleaned_data it just adds it to a list of errors. If we replaced "abc" with "ab" it will correctly fill out form.cleaned_data